// ngx-onsenui/directives/ons-navigator.js
import { Injector, ReflectiveInjector, Directive, ElementRef, ComponentFactoryResolver, Input, ViewContainerRef, NgZone } from '@angular/core';
import * as ons from 'onsenui';
import { Params } from '../ons/params';
var NavigatorPage = (function () {
    function NavigatorPage(elementRef, destroy, animator, params) {
        if (elementRef === void 0) { elementRef = null; }
        if (destroy === void 0) { destroy = function () { }; }
        if (animator === void 0) { animator = null; }
        if (params === void 0) { params = new Params(); }
        this.elementRef = elementRef;
        this.destroy = destroy;
        this.animator = animator;
        this.params = params;
    }
    return NavigatorPage;
}());
export { NavigatorPage };
/**
 * @element ons-navigator
 * @directive OnsNavigator
 * @selector ons-navigator
 * @codpen KgvAvx
 * @description
 *   [en]Angular directive for `<ons-navigator>` component.[/en]
 *   [ja]`<ons-navigator>`要素のAngularディレクティブです。[/ja]
 * @example
 *   @Component({
 *     selector: 'ons-page',
 *     template: `
 *       <ons-toolbar>
 *         <div class="center">Page</div>
 *       </ons-toolbar>
 *       <div class="content">...</div>
 *     `
 *   })
 *   class DefaultPageComponent { }
 *
 *   @Component({
 *     selector: 'navigator-app',
 *     template: `
 *     <ons-navigator [page]="page"></ons-navigator>
 *     `
 *   })
 *   export class AppComponent {
 *     page = DefaultPageComponent
 *   }
 */
var OnsNavigator = (function () {
    function OnsNavigator(_elementRef, _resolver, _viewContainer, _injector, _zone) {
        this._elementRef = _elementRef;
        this._resolver = _resolver;
        this._viewContainer = _viewContainer;
        this._injector = _injector;
        this._zone = _zone;
        this._lastPageLoader = this.element.pageLoader;
        this.element.pageLoader = this._createPageLoader();
    }
    Object.defineProperty(OnsNavigator.prototype, "pageComponentType", {
        /**
         * @input page
         * @type {Type<any>}
         * @desc
         *   [en]Type of the page component.[/en]
         *   [ja]ページコンポーネントのクラスを指定します。[/ja]
         */
        set: function (page) {
            this.element.page = page;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OnsNavigator.prototype, "element", {
        get: function () {
            return this._elementRef.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OnsNavigator.prototype, "nativeElement", {
        get: function () {
            return this._elementRef.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    OnsNavigator.prototype._createPageLoader = function () {
        var _this = this;
        var componentRefMap = new WeakMap();
        return new ons.PageLoader(function (_a, done) {
            var page = _a.page, parent = _a.parent, params = _a.params;
            _this._zone.run(function () {
                var pageParams = new Params(params || {});
                var injector = ReflectiveInjector.resolveAndCreate([
                    { provide: Params, useValue: pageParams },
                    { provide: OnsNavigator, useValue: _this }
                ], _this._injector);
                var factory = _this._resolver.resolveComponentFactory(page);
                var selector = 'ons-navigator';
                var pageComponentRef = factory.create(injector, undefined);
                _this._viewContainer.insert(pageComponentRef.hostView);
                var pageElement = pageComponentRef.location.nativeElement;
                componentRefMap.set(pageElement, pageComponentRef);
                _this.element.appendChild(pageElement); // dirty fix to insert in correct position
                done(pageElement);
            });
        }, function (element) {
            var componentRef = componentRefMap.get(element);
            if (componentRef) {
                componentRef.destroy();
                componentRefMap.delete(element);
            }
        });
    };
    OnsNavigator.prototype.ngOnDestroy = function () {
        this.element.pageLoader = this._lastPageLoader;
    };
    return OnsNavigator;
}());
export { OnsNavigator };
OnsNavigator.decorators = [
    { type: Directive, args: [{
                selector: 'ons-navigator'
            },] },
];
/** @nocollapse */
OnsNavigator.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: ComponentFactoryResolver, },
    { type: ViewContainerRef, },
    { type: Injector, },
    { type: NgZone, },
]; };
OnsNavigator.propDecorators = {
    'pageComponentType': [{ type: Input, args: ['page',] },],
};
